# Chan Kai Yang
# 7372711

## Importing Relevant Libraries

In [1]:
import pandas as pd
import numpy as np
import random
import sys
from random import randrange
from math import sqrt
from math import pi
from math import exp
from math import isnan

## Pre-Processing
### Reading data file

In [2]:
# read csv data file
#Add in headers for the data frame
#Ensure sep =', ' as there is a space before each feature
df = pd.read_csv("adult.data", sep=", ", engine='python',
                  names=["age", "workclass", "fnlwgt", "education", "education-num", "marital-status", "occupation", "race", "relationship", "sex", "capital-gain", "capital-loss", "hours-per-week", "native-country", "income"])

df.head(30)

#Check format of csv file
#import csv

#with open("./adult.data", 'r') as file:
#  csvreader = csv.reader(file)
#  for row in csvreader:
#    print(row)

age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
14   40           Private  121772     Assoc-voc             11   
15   34           Private  245487       7th-8th              4   
16   25  Self-emp-not-inc  176756       HS-grad              9   
17   32           Private  186824       HS-grad              9   
18   38           Private   28887          11th              7   
19   43  Self-emp-not-inc  292175       Masters             14   
20   40           Private  193524     Doctorate             16   
21   54           Private  302146       HS-grad              9   
22   35       Federal-gov   76845           9th              5   
23   43           Private  117037          11th              7   
24   59           Private  109015       HS-grad              9   
25   56         Local-gov  216851     Bachelors             13   
26   19           Private  168294       HS-grad              9   
27   54                 ?  180211  Some-college             10   
28   39           Private  367260       HS-grad              9   
29   49           Private  193366       HS-grad              9   

           marital-status         occupation           race  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-married     Prof-specialty  Not-in-family   
9      Married-civ-spouse    Exec-managerial        Husband   
10     Married-civ-spouse    Exec-managerial        Husband   
11     Married-civ-spouse     Prof-specialty        Husband   
12          Never-married       Adm-clerical      Own-child   
13          Never-married              Sales  Not-in-family   
14     Married-civ-spouse       Craft-repair        Husband   
15     Married-civ-spouse   Transport-moving        Husband   
16          Never-married    Farming-fishing      Own-child   
17          Never-married  Machine-op-inspct      Unmarried   
18     Married-civ-spouse              Sales        Husband   
19               Divorced    Exec-managerial      Unmarried   
20     Married-civ-spouse     Prof-specialty        Husband   
21              Separated      Other-service      Unmarried   
22     Married-civ-spouse    Farming-fishing        Husband   
23     Married-civ-spouse   Transport-moving        Husband   
24               Divorced       Tech-support      Unmarried   
25     Married-civ-spouse       Tech-support        Husband   
26          Never-married       Craft-repair      Own-child   
27     Married-civ-spouse                  ?        Husband   
28               Divorced    Exec-managerial  Not-in-family   
29     Married-civ-spouse       Craft-repair        Husband   

   

In [3]:
#Check for null values or missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   race            32561 non-null  object
 8   relationship    32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [4]:
#Notice workclass and occupation has a '?' value on the same row (row 27)
#Notice native-country has a '?' value on row 14
#Drop all rows with '?' values as they are not in the attribute information
#irrelevant as we cannot get any information about workclass or occupation from ?

#Loop through all columns in dataframe and remove the rows with '?'
for column in df:
    df.drop(df.index[df[column] == '?'], inplace=True)

#Check that all rows with '?' have been removed
df.head(30)

#Row 27 is removed
#Row 14 is removed

age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
15   34           Private  245487       7th-8th              4   
16   25  Self-emp-not-inc  176756       HS-grad              9   
17   32           Private  186824       HS-grad              9   
18   38           Private   28887          11th              7   
19   43  Self-emp-not-inc  292175       Masters             14   
20   40           Private  193524     Doctorate             16   
21   54           Private  302146       HS-grad              9   
22   35       Federal-gov   76845           9th              5   
23   43           Private  117037          11th              7   
24   59           Private  109015       HS-grad              9   
25   56         Local-gov  216851     Bachelors             13   
26   19           Private  168294       HS-grad              9   
28   39           Private  367260       HS-grad              9   
29   49           Private  193366       HS-grad              9   
30   23         Local-gov  190709    Assoc-acdm             12   
31   20           Private  266015  Some-college             10   

           marital-status         occupation           race  \
0           Never-married       Adm-clerical  Not-in-family   
1      Married-civ-spouse    Exec-managerial        Husband   
2                Divorced  Handlers-cleaners  Not-in-family   
3      Married-civ-spouse  Handlers-cleaners        Husband   
4      Married-civ-spouse     Prof-specialty           Wife   
5      Married-civ-spouse    Exec-managerial           Wife   
6   Married-spouse-absent      Other-service  Not-in-family   
7      Married-civ-spouse    Exec-managerial        Husband   
8           Never-married     Prof-specialty  Not-in-family   
9      Married-civ-spouse    Exec-managerial        Husband   
10     Married-civ-spouse    Exec-managerial        Husband   
11     Married-civ-spouse     Prof-specialty        Husband   
12          Never-married       Adm-clerical      Own-child   
13          Never-married              Sales  Not-in-family   
15     Married-civ-spouse   Transport-moving        Husband   
16          Never-married    Farming-fishing      Own-child   
17          Never-married  Machine-op-inspct      Unmarried   
18     Married-civ-spouse              Sales        Husband   
19               Divorced    Exec-managerial      Unmarried   
20     Married-civ-spouse     Prof-specialty        Husband   
21              Separated      Other-service      Unmarried   
22     Married-civ-spouse    Farming-fishing        Husband   
23     Married-civ-spouse   Transport-moving        Husband   
24               Divorced       Tech-support      Unmarried   
25     Married-civ-spouse       Tech-support        Husband   
26          Never-married       Craft-repair      Own-child   
28               Divorced    Exec-managerial  Not-in-family   
29     Married-civ-spouse       Craft-repair        Husband   
30          Never-married    Protective-serv  Not-in-family   
31          Never-married              Sales      Own-child   

   

In [5]:
#Check that all rows with '?' have been removed
df.info()

#Around 2k rows removed

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   workclass       30162 non-null  object
 2   fnlwgt          30162 non-null  int64 
 3   education       30162 non-null  object
 4   education-num   30162 non-null  int64 
 5   marital-status  30162 non-null  object
 6   occupation      30162 non-null  object
 7   race            30162 non-null  object
 8   relationship    30162 non-null  object
 9   sex             30162 non-null  object
 10  capital-gain    30162 non-null  int64 
 11  capital-loss    30162 non-null  int64 
 12  hours-per-week  30162 non-null  int64 
 13  native-country  30162 non-null  object
 14  income          30162 non-null  object
dtypes: int64(6), object(9)
memory usage: 4.7+ MB


In [6]:
#After further analysis, education-num is the pre-encoded values of education, hence continuous value
# Hence, we can drop education for processing

# drop 'education' from the DataFrame
df = df.drop('education', axis=1)

In [7]:
# Encode workclass
newdf = pd.Series(df['workclass'], dtype = "category")
df['workclass'] = newdf.cat.codes

# Encode marital-status
newdf = pd.Series(df['marital-status'], dtype = "category")
df['marital-status'] = newdf.cat.codes

# Encode occupation
newdf = pd.Series(df['occupation'], dtype = "category")
df['occupation'] = newdf.cat.codes

# Encode race
newdf = pd.Series(df['race'], dtype = "category")
df['race'] = newdf.cat.codes

# Encode race
newdf = pd.Series(df['relationship'], dtype = "category")
df['relationship'] = newdf.cat.codes

# Encode sex
newdf = pd.Series(df['sex'], dtype = "category")
df['sex'] = newdf.cat.codes

# Encode native-country
newdf = pd.Series(df['native-country'], dtype = "category")
df['native-country'] = newdf.cat.codes

# Encode income
newdf = pd.Series(df['income'], dtype = "category")
df['income'] = newdf.cat.codes

df.head()

age  workclass  fnlwgt  education-num  marital-status  occupation  race  \
0   39          5   77516             13               4           0     1   
1   50          4   83311             13               2           3     0   
2   38          2  215646              9               0           5     1   
3   53          2  234721              7               2           5     0   
4   28          2  338409             13               2           9     5   

   relationship  sex  capital-gain  capital-loss  hours-per-week  \
0             4    1          2174             0              40   
1             4    1             0             0              13   
2             4    1             0             0              40   
3             2    1             0             0              40   
4             2    0             0             0              40   

   native-country  income  
0              38       0  
1              38       0  
2              38       0  
3              38       0  
4               4       0

### Stratify Dataset into Training and Testing Set Function

In [8]:
# Stratifying the data into training data set and testing data set
# The testSize is specified as percentage for test set

def trainTestSplit(data, testSize):
    # testSize * no. of rows in df = specified numbers of rows 
    if isinstance(testSize, float):
        testSize = round(testSize * len(data))
        
    # Getting the index of the data from the dataset
    dataIndex = data.index.tolist()
    
    # Generate random indexes to split the dataset (stratified sampling)
    testIndex = random.sample(population = dataIndex, k = testSize)
    
    # Allocate the testing set according to the random indexes (testIndex)
    testSet = data.loc[testIndex]
    
    # Dropping the testset rows from the original data set, and make the remaining rows as the training set
    trainSet = data.drop(testIndex)
    
    # Return the train and test set data
    return trainSet, testSet

### Gaussian Naive Bayes Function

In [9]:
# Calculate the Gaussian probability distribution for continuous features
# To prevent numerical underflow, when the exponent underflows is out of Pythons' float precision, set
# exponent to the smallest possible float value by the system
# To prevent the zero frequency/count problem, the function catches occurences of 0/null standard deviation
# and returns the probability of 1/number of observations

def calculateGaussianProbability(x, mean, stdev, totalRows):
    # If the standard deviation is zero or not a number, return a small value to avoid division by zero
    if stdev == 0 or isnan(stdev):
        return 1/totalRows # if stdev is 0, return the probability of "Add one count"
    
    # Calculate the exponent for the Gaussian distribution
    exponent = exp(-((x - mean)**2 / (2 * stdev**2 )))
    
    # If the exponent is zero, set it to the smallest float value to avoid division by zero and numerical underflow
    if exponent == 0:
        exponent = sys.float_info.min
    
    # Calculate the Gaussian probability using the mean, standard deviation, and exponent
    return (1 / (sqrt(2 * pi) * stdev)) * exponent

### Function to calculate probability for categorical features

In [10]:
# Calculate probability for categorical features
# To prevent the zero frequency/count problem, the function catches the occurences of 0/null standard deviation

def calculateProbability(x, X1, count1, X2, count2, classCount, totalRows):
    # If the feature value is X1
    if x == X1:
        # If the count is zero, return a small value to avoid division by zero
        if count1/classCount == 0:
            return 1/totalRows
        
        return count1/classCount
    else:
        # If the count is zero, return a small value to avoid division by zero
        if count2/classCount == 0:
            return 1/totalRows
        
        return count2/classCount
    

### Function to calculate class probability

In [11]:
# Calculate the probabilities of predicting each class for a given row:
# For continuous features use Gaussian probability function
# For categorical feature calculate_probability function

def calculateClassProbabilities(summaries, row):
    
    # Get the total number of rows of the dataset
    # Sum up all the counts of each label class
    totalRows = sum([summaries[label][0][2] for label in summaries])
    
    # Create a dictionary to store the probability for each class label for a given row
    probabilities = dict()
    
    # Get the class value: classValue
    # Get the summaries for each class: classSummaries
    for classValue, classSummaries in summaries.items():
        
        # Get the probability of each label class e.g., If class label 1 has a length of 12
        # and length of dataset is 12345 then this probability is 12/12345
        probabilities[classValue] = summaries[classValue][0][2]/float(totalRows)
        
        #looping through each summaries
        for i in range(len(classSummaries)):
            # if categorical feature
            if len(classSummaries[i]) > 3:
                X1, count1, X2, count2, classCount = classSummaries[i]
                probabilities[classValue] = \
                probabilities[classValue] * calculateProbability(row[i], X1, count1, X2, count2, classCount, totalRows)
            # if continuous feature
            else:
                mean, stdev, _ = classSummaries[i]
                probabilities[classValue] = \
                probabilities[classValue] * calculateGaussianProbability(row[i], mean, stdev, totalRows)
                
    return probabilities

### Predict Class of Row

In [12]:
# Predict the class for a given row
def predictClass(summaries, row):
    # Storing the probabilities
    probabilities = calculateClassProbabilities(summaries, row)
    
    # Initializing the variables to keep track of the best class label and its corresponding probability
    bestLabel, bestProb = None, -1
    
    # Looping through to find the best label with the best probability
    for classVal, probability in probabilities.items():
        if bestLabel is None or probability > bestProb:
            bestProb = probability
            bestLabel = classVal
    return bestLabel

### Function to get a summary dictionary of dataset

In [13]:
def summarizeDataset(data):
    # Creating a empty dict to store the target class
    summaries = {}
    # For loop to iterate through the target column that are unique
    for i in data.iloc[:,-1].unique():
        # listing the feature variable of each unique class
        feature = []
        # Iterate through all the feature variable except the target column
        for j in range(len(data.columns)-1):
            # Storing the size of unique values in a variable
            uniqueValueSize = len(data.iloc[:,j].unique())
            
            # If categorical feature
            if(uniqueValueSize < 5):
                aList = list()
                # Creating a subset for each class
                df = data[data.iloc[:,-1] == i]
                
                # Iterate over each unique value in the categorical feature
                for k in data.iloc[:,j].unique():
                    # storing the unique value and counting the val
                    aList.append(k)
                    aList.append(len(df[df.iloc[:,j] == k]))
                feature.append(tuple([aList[0], aList[1], aList[2], aList[3], len(data[data.iloc[:,-1] == i])]))
            # else continuous feature
            else:
                feature.append((data[data.iloc[:,-1] == i].mean(axis = 0)[j], \
                                data[data.iloc[:,-1] == i].std(axis = 0)[j], len(data[data.iloc[:,-1] == i])))
        # storing the summaries     
        summaries[i] = feature
    return summaries

In [14]:
# Naive Bayes Classifier Algorithm to predict the train set and the test set
def naiveBayesian(trainSet, testSet):
    # Method to get the summary
    summary = summarizeDataset(trainSet)
    
    # Creating an empty list
    predictions = list()
    
    # looping through the test set values 
    for row in testSet.values:
        output = predictClass(summary, row)
        predictions.append(output)
        
    return(predictions)

### Function to determine regression metrics

In [15]:
# Calculated as:
# check for equality of predicted value and labels in test_set
# calculates the sum of correct prediction
# divides the sum by length of test_set
    
def accuracy(predictions, dataSet):
    yTest = list(dataSet.iloc[:,-1])
    correctCount = 0
    sumError = 0.0
    rsmeError = 0.0
    
    # Loop through the predictions and actual values and update counters
    for i in range(len(yTest)):
        if predictions[i] == yTest[i]:
            correctCount += 1
        sumError += abs(predictions[i] - yTest[i])
        predictionError = abs(predictions[i] - yTest[i])
        rsmeError = (predictionError**2)
        
    print(f'Number of exact matches in predictions: {correctCount}/{len(yTest)}')
    
    # return the percentage of correctly predicted values
    return (round(correctCount/len(dataSet)*100,3))

### Split dataset into training (70%) and testing set (30%)

In [16]:
# Split the dataset into training (70%) and testing set (30%)
train_set, test_set = trainTestSplit(df, 0.3)

#Check count of training set
train_set.describe()

age     workclass        fnlwgt  education-num  \
count  21113.000000  21113.000000  2.111300e+04   21113.000000   
mean      38.468763      2.199214  1.896145e+05      10.111779   
std       13.117173      0.952553  1.055920e+05       2.566012   
min       17.000000      0.000000  1.487800e+04       1.000000   
25%       28.000000      2.000000  1.173130e+05       9.000000   
50%       37.000000      2.000000  1.783190e+05      10.000000   
75%       47.000000      2.000000  2.369380e+05      13.000000   
max       90.000000      6.000000  1.484705e+06      16.000000   

       marital-status    occupation          race  relationship           sex  \
count    21113.000000  21113.000000  21113.000000  21113.000000  21113.000000   
mean         2.571165      5.941458      1.424430      3.681429      0.672713   
std          1.501473      4.034087      1.604419      0.832372      0.469234   
min          0.000000      0.000000      0.000000      0.000000      0.000000   
25%          2.000000      2.000000      0.000000      4.000000      0.000000   
50%          2.000000      6.000000      1.000000      4.000000      1.000000   
75%          4.000000      9.000000      3.000000      4.000000      1.000000   
max          6.000000     13.000000      5.000000      4.000000      1.000000   

       capital-gain  capital-loss  hours-per-week  native-country  \
count  21113.000000  21113.000000    21113.000000    21113.000000   
mean    1038.931938     91.049922       40.852887       36.366173   
std     7057.367796    410.110492       11.897223        6.141798   
min        0.000000      0.000000        1.000000        0.000000   
25%        0.000000      0.000000       40.000000       38.000000   
50%        0.000000      0.000000       40.000000       38.000000   
75%        0.000000      0.000000       45.000000       38.000000   
max    99999.000000   3900.000000       99.000000       40.000000   

             income  
count  21113.000000  
mean       0.248615  
std        0.432220  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [17]:
#Check count of testing set
test_set.describe()

age    workclass        fnlwgt  education-num  marital-status  \
count  9049.000000  9049.000000  9.049000e+03    9049.000000     9049.000000   
mean     38.365897     2.199580  1.902123e+05      10.143552        2.601061   
std      13.175831     0.957171  1.057997e+05       2.512228        1.489791   
min      17.000000     0.000000  1.376900e+04       1.000000        0.000000   
25%      28.000000     2.000000  1.188760e+05       9.000000        2.000000   
50%      37.000000     2.000000  1.786860e+05      10.000000        2.000000   
75%      47.000000     2.000000  2.389800e+05      13.000000        4.000000   
max      90.000000     6.000000  1.268339e+06      16.000000        6.000000   

        occupation         race  relationship          sex  capital-gain  \
count  9049.000000  9049.000000   9049.000000  9049.000000   9049.000000   
mean      6.002763     1.404133      3.672008     0.682617   1215.843850   
std       4.018895     1.594125      0.840148     0.465484   8161.843318   
min       0.000000     0.000000      0.000000     0.000000      0.000000   
25%       2.000000     0.000000      4.000000     0.000000      0.000000   
50%       6.000000     1.000000      4.000000     1.000000      0.000000   
75%       9.000000     3.000000      4.000000     1.000000      0.000000   
max      13.000000     5.000000      4.000000     1.000000  99999.000000   

       capital-loss  hours-per-week  native-country       income  
count   9049.000000     9049.000000     9049.000000  9049.000000  
mean      82.125539       41.114046       36.420820     0.249641  
std      390.352125       12.169596        6.019691     0.432829  
min        0.000000        1.000000        0.000000     0.000000  
25%        0.000000       40.000000       38.000000     0.000000  
50%        0.000000       40.000000       38.000000     0.000000  
75%        0.000000       45.000000       38.000000     0.000000  
max     4356.000000       99.000000       40.000000     1.000000

### Test Naive Bayes Model

In [18]:
# Test the model on training set
nb_train_pred = naiveBayesian(train_set, train_set)
print('Accuracy of prediction for training set:', accuracy(nb_train_pred, train_set))

Number of exact matches in predictions: 16736/21113
Accuracy of prediction for training set: 79.269


Number of exact matches in predictions: 16736/21113: The Naive Bayes Classifier correctly predicted the class of 16736 out of 21113 instances in the training set.

Accuracy of prediction for training set: 79.269: An accuracy of 79.269 means that the model correctly predicted 79.269% of the samples in the training dataset.

Hence, the Naive Bayes model is pretty accurate in predicting income.

In [19]:
# Test the model on testing set
nb_test_pred = naiveBayesian(train_set, test_set)
print('Accuracy of prediction for testing set:', accuracy(nb_test_pred, test_set))

Number of exact matches in predictions: 7141/9049
Accuracy of prediction for testing set: 78.915


Number of exact matches in predictions: 7141/9049: The Naive Bayes Classifier correctly predicted the class of 7141 out of 9049 instances in the testing set.

Accuracy of prediction for training set: 78.915: An accuracy of 78.915 means that the model correctly predicted 78.915% of the samples in the testing dataset.

Hence, the Naive Bayes model is pretty accurate in predicting income.